### PR TITLE
items: add new fields to inventory list

### LINKED ISSUE #2704: Add some fields to the Inventory list
**How it works**

In the inventory list, some fields are missing to help librarians in the collection management.

**Improvement suggestion**

Please add the following fields :  
From the document records 
- Identifiers : ISBN and ISSN
- Series statement : _text (so we can see all the parts and numbering)
- Provision Activity : Agent (in the Statement)
- Provision Activity : startDate and endDate
- Edition statement : _text
- Masked : Yes/No
- Local fields of the organisation

From the item records and loans
- Creation date of the item
- Temporary circulation category : expiry date
- All the notes type (notes.type:"acquisition_note" ; "binding_note" ; "condition_note" ; "patrimonial_note" ; "provenance_note"). [in addition to the one that are currently in the inventory]
- Masked : Yes/No
- PAC code
- Number of renewals
- Local fields

Check the field called "created" in the current CSV export
- if it is the creation of the item, put the field in the beginning of the columns
- if it is another creation date, rename the field

See current bug at exporting loan count: https://github.com/rero/rero-ils/issues/2793

Nice to have : Be able to choose the fields we need when exporting

### LINKED ISSUE #2793: Inventory list export: some columns counts only use active loans
### How it works

When exporting a CSV from an inventory list (pro UI), some columns only count the values from currently active loans (`ITEM_AT_DESK`, `ITEM_ON_LOAN`, `ITEM_IN_TRANSIT_FOR_PICKUP`, `ITEM_IN_TRANSIT_TO_HOUSE`) for an item where they should be more global:

* `loans_count` column is always binary (either 0 or 1). It counts the number of active loans. This is not needed for users; they can get this information from the `status` column.
* `last_transaction_date` only displays something if there is an active loan for the item. Furthermore it displays the last date of the document `operation log`. This should be from the item/loan operation log. Librarians want to see when the item was last circulated, whether the loan is still active or not. If an item has been `ON_SHELF` for six months, they want to know.

### Improvement suggestion

* `loans_count` : in this column we want to display the number of loans for an item, i.e. the number of checkout transactions this item has had in its entire lifetime (like for the stats module). 
* `last_transaction_date` : for the values to be coherent with this label, this should show the latest transaction date whether the item still has an active loan or not. 

### Nice to have

Add one more column in the export (after `loans_count`) that also displays the number of renewal transaction that the item has had in its lifetime. Call this (`renewals_count`).

### PR DESCRIPTION
**Pull request by BadrAly**
Adds fields from item and document resources to the inventory list.

* Fixes #2704
* Fixes #2793

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

# How to Test:

- run the `Inventory list` report  for a given instance.
- the fields extracted from the `document` record are now prefixed with the string `document_`
- the fields extracted from the `item` record are now prefixed with the string `item_`
- the `local_fields` are now extracted for both `documents` and `items`
- all the fields mentioned in the issue https://github.com/rero/rero-ils/issues/2704 are added
- loan information are correctly extracted according to https://github.com/rero/rero-ils/issues/2793

### CODE REVIEW

File: rero_ils/modules/items/serializers/collector.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        :return list of loans.
+        """
+        states = \
+            current_app.config['CIRCULATION_STATES_LOAN_ACTIVE']
+        loan_search = LoansSearch() \
+            .filter('terms', state=states) \
+            .filter('terms', item_pid__value=item_pids) \
+            .source(['pid', 'item_pid.value', 'start_date',
+                    'end_date', 'state', '_created'])
+        agg = A('terms', field='item_pid.value', size=cls.chunk_size)
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/items/serializers/collector.py]: I'm re-writing this part, will get in touch with Laurent...

File: rero_ils/modules/items/serializers/csv.py [hunk 2]
```diff
@@ -18,277 +18,77 @@
...
+            for hit in batch_results:
+                csv_data = Collecter.get_item_data(hit, item_types_map)
+                Collecter.append_library_data(csv_data, libraries_map)
+                Collecter.append_location_data(csv_data, locations_map)
+                Collecter.append_item_local_fields(csv_data)
+                Collecter.append_document_data(csv_data, documents)
+                Collecter.append_document_local_fields(csv_data)
+                # update csv data with loan
+                csv_data['item_checkouts_count'] = checkouts.get(hit.pid, 0)
+                csv_data['item_renewals_count'] = renewals.get(hit.pid, 0)
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/items/serializers/csv.py:75]: @lauren-d: I am adding the missing fields...

**Review by rerowep — CHANGES_REQUESTED** [review]

File: rero_ils/modules/items/serializers/collector.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+        results = renewals_query[:cls.chunk_size].execute()
+        renewals = {
+            result.key: result.doc_count
+            for result in results.aggregations.loans_count.buckets}
+
+        last_transaction_dates = {}
+        date_query = RecordsSearch(index=LoanOperationLog.index_name)\
+            .filter('terms', loan__item__pid=item_pids) \
+            .sort({'_created': {"order": "desc"}})\
```
[the remark this answers is not in the document]
Comment (lauren-d) [reply rero_ils/modules/items/serializers/collector.py:385]: probably not needed with execute(). must be tested.

File: rero_ils/modules/items/serializers/collector.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+    @classmethod
+    def get_loans_by_item_pids(cls, item_pids=None):
+        """Get loans for the given item pid list.
+
+        :param item_pids: item pids.
+        :return list of checkouts, renewals, last_transaction_dates,
+                last_checkouts.
+        """
+        checkouts_query = RecordsSearch(index=LoanOperationLog.index_name)\
+            .filter('term', loan__trigger=ItemCirculationAction.CHECKOUT)\
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector.py:361]: Mmmm...It seems that `loan.trigger` isn't mapped into ES. It should be no ?
If it's not mapped into ES, it should be analyze as "text" with a raw (keyword) children attribute (@jma : is it correct ?)
So either add this field into ES_mapping (... and create an alembic script to reindex all alive loans), either use `loan__trigger__raw`. (I vote for the first solution !)

File: rero_ils/modules/items/serializers/collector.py [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+            .filter('terms', loan__item__pid=item_pids)\
+            .source(['loan.item.pid'])
```
Comment (zannkukai) [inline rero_ils/modules/items/serializers/collector.py:363]: you can add an `.extra(rows=0)` if you didn't need to check result hits.

**Review by zannkukai — COMMENTED** [review]
`get_loans_by_item_pids` is comming

File: rero_ils/modules/items/serializers/collector.py [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+        org_pid = csv_data['item_org_pid']
+        if local_fields := LocalField.get_local_fields_by_resource(
+                resource_type, resource_pid, organisation_pid=org_pid):
+            for field, num in itertools.product(local_fields, range(1, 11)):
+                field_name = f'{resource_type}_local_field_{num}'
+                if field_data := field.get('fields', {}).get(f'field_{num}'):
+                    csv_data[field_name] = ' '.join(filter(None, field_data))
+
+    @classmethod
+    def get_loans_by_item_pids(cls, item_pids=None):
```
Comment (lauren-d) [inline rero_ils/modules/items/serializers/collector.py:343]: could you put more comments in this method?

**Pull request closed without merging**